mav_trajectory_generation/include/mav_trajectory_generation/impl/polynomial_optimization_nonlinear_impl.h

```diff
@@ -156,15 +170,289 @@ int PolynomialOptimizationNonLinear<_N>::optimizeTime() {
+      // Now the same with an increased segment time
+      // Calculate cost with higher segment time
+      segment_times_bigger = segment_times;
+      // Deduct h*(-1/(m-2)) according to paper Mellinger "Minimum snap traject
+      // generation and control for quadrotors"
+      double const_traj_time_corr = increment_time/(n_segments-1.0);
+      for (int i = 0; i < segment_times_bigger.size(); ++i) {
+        if (i==n) {
+          segment_times_bigger[i] += increment_time;
+        } else {
+          segment_times_bigger[i] -= const_traj_time_corr;
+        }
+      }
+
+      // TODO: add case if segment_time is at threshold 0.1s
+      // 1) How many segments > 0.1s
+      // 2) trajectory time correction only on those
+      // for (int j = 0; j < segment_times_bigger.size(); ++j) {
+      //   double thresh_corr = 0.0;
+      //   if (segment_times_bigger[j] < 0.1) {
+      //     thresh_corr = 0.1-segment_times_bigger[j];
+      //   }
+      // }
+
+      // Check and make sure if segment times are all > 0.1
+      for (double& t : segment_times_bigger) {
+        t = t <= kOptimizationTimeLowerBound ? kOptimizationTimeLowerBound : t;
+      }
+
+      // Update the segment times. This changes the polynomial coefficients.
+      poly_opt_.updateSegmentTimes(segment_times_bigger);
+      poly_opt_.solveLinear();
+
+      // Calculate cost and gradient with new segment time
+      const double J_d_bigger = poly_opt_.computeCost();
+      const double dJd_dt = (J_d_bigger - J_d) / increment_time;
+
+      // Calculate the gradient
+      gradients->at(n) = w_d*dJd_dt;
+    }
+
+    // Set again the original segment times from before calculating the
+    // numerical gradient
+    poly_opt_.updateSegmentTimes(segment_times);
+    poly_opt_.solveLinear();
+  }
+
+  // Compute cost without gradient
+  return w_d*J_d;
+}
+
+template <int _N>
+void PolynomialOptimizationNonLinear<_N>::scaleSegmentTimesWithViolation(
+        Eigen::VectorXd* segment_times) {
+  // Get trajectory
+  Trajectory traj;
+  poly_opt_.getTrajectory(&traj);
+
+  // Evaluate min/max extrema
+  Extremum v_min_actual, v_max_actual, a_min_actual, a_max_actual;
+  std::vector<int> dimensions = {0, 1, 2}; // Evaluate dimensions in x, y and z
+  traj.computeMinMaxMagnitude(derivative_order::VELOCITY, dimensions,
+                              &v_min_actual, &v_max_actual);
+  traj.computeMinMaxMagnitude(derivative_order::ACCELERATION, dimensions,
+                              &a_min_actual, &a_max_actual);
+
+  // Get constraints
+  double v_max = 0.0;
+  double a_max = 0.0;
+  for (const auto& constraint : inequality_constraints_) {
+    if (constraint->derivative == derivative_order::VELOCITY) {
+      v_max = constraint->value;
+    } else if (constraint->derivative == derivative_order::ACCELERATION) {
+      a_max = constraint->value;
+    }
+  }
+
+  // Evaluate constraint/bound violation
+  double abs_violation_v = v_max_actual.value - v_max;
```

Comment:
Is this possible? I have assumed that v_max and v_max_actual is always the absolute magnitude and thus always positive?

Reply:
Is it even possible to set a minimum constraint?
The struct ConstraintData only takes one constraint per derivative where we dont distinguish between min and max (assumed to be max since only method addMaximumMagnitudeConstraint exists).

```
  struct ConstraintData {
    PolynomialOptimizationNonLinear<N>* this_object;
    int derivative;
    double value;
  };
```

The code currently does not support adding MinimumMagnitudeConstraints, correct?

Reply:
Well the call above has:
```
+  traj.computeMinMaxMagnitude(derivative_order::ACCELERATION, dimensions,
+                              &a_min_actual, &a_max_actual);
```
It makes sense that magnitude should be absolute, yes, so max magnitude should be the absolute max (and min magnitude should probably be 0).
 Could you quickly check the value of a_min_actual and make sure it's positive and the function name matches what it's actually doing? If so then yes, this way is correct. :)

Reply:
computeMinMaxMagnitude only computes the minimum magnitude --> which is always 0.0.
However, we do not have the possibility of setting the minimum magnitude, eg setting a minimum velocity (which whould be interesting but not really necessary for now)